In [2]:
import os
from dotenv import load_dotenv, find_dotenv
_ = load_dotenv(find_dotenv())
openai_api_key = os.environ["OPENAI_API_KEY"]

## With Chat Model

In [3]:
from langchain_openai import ChatOpenAI

chat_model = ChatOpenAI(model="gpt-3.5-turbo-0125")

In [4]:
messages = [
    ("system", "You are an historian expert in the history of the Roman Empire."),
    ("human", "Tell me about the fall of the Roman Empire.")
]

response = chat_model.invoke(messages)
print(response.content)

The fall of the Western Roman Empire is a complex and multifaceted event that took place over several centuries, but is traditionally dated to 476 AD when the last Roman emperor of the West, Romulus Augustulus, was deposed by the Germanic chieftain Odoacer. There are many factors that contributed to the decline and fall of the Western Roman Empire, including political, economic, social, and military causes.

One of the key factors leading to the fall of the Western Roman Empire was political instability. After the death of Emperor Marcus Aurelius in 180 AD, the empire entered a period known as the Crisis of the Third Century, marked by frequent changes in leadership, civil wars, and usurpations. This weakened the authority of the central government and made the empire more vulnerable to external threats.

Economic troubles also played a significant role in the decline of the Western Roman Empire. The empire faced inflation, debasement of the currency, heavy taxation, and economic dispa

In [7]:
messages = [
    ("system", "You are a Geopolitics expert."),
    ("human", "Tell me about the tarriff imposed by US on China in 2025.")
]

response = chat_model.invoke(messages)
print(response.content)

I'm sorry, but I cannot provide real-time or future information as my training data only goes up until September 2021. However, I can tell you about the general context of the trade tensions and tariffs between the US and China up to that point.

The trade war between the US and China started in 2018 when the US imposed tariffs on billions of dollars worth of Chinese goods, citing issues such as intellectual property theft and unfair trade practices. In retaliation, China also imposed tariffs on US goods. This tit-for-tat escalation of tariffs had significant impacts on both economies and the global economy.

By 2025, it's possible that the trade tensions between the US and China may have continued, resulting in further tariffs being imposed on each other's goods. These tariffs can have wide-ranging effects on industries, businesses, and consumers in both countries, as well as implications for the global economy.

For the most up-to-date information on any tariffs imposed by the US on 

## Data Loader

https://python.langchain.com/docs/integrations/document_loaders/

In [11]:
# txt data loading
from langchain_community.document_loaders import TextLoader

loader = TextLoader("data/llms.txt")
loaded_data = loader.load()
loaded_data

[Document(metadata={'source': 'data/llms.txt'}, page_content="# About\nName: Fyno\nDescription: Fyno is a modern infrastructure for product and engineering teams to build and manage their notification or communications service with minimum effort.  \nURL: https://www.fyno.io/blog\n\n# Navigation Menu\n- Blogs: https://fyno.io/blog\n- Talk to us: https://fyno.io/schedule-demo\n- Try for Free: https://app.fyno.io/login\n\n# Blog Posts\n## Why 100% message deliverability matters in banking (and how to achieve it)\nAuthor: Karthik Pasupathy\nPublished: 2025-04-04\nCategory: Articles\nTags: Customer communication\nURL: https://www.fyno.io/blog/why-100percent-message-deliverability-matters-in-banking-and-how-to-achieve-it-cm92ajaan000fi9pa4ylyzykk\n\nIn banking, a simple missed notification can have serious consequences. Imagine a customer unable to complete a transaction because an OTP never arrived, or missing a fraud alert that could have prevented unauthorized access to their account. Th

In [13]:
# csv data loading
from langchain_community.document_loaders.csv_loader import CSVLoader

loader = CSVLoader("data/corpus_metadata.csv")
loaded_data = loader.load()
loaded_data

[Document(metadata={'source': 'data/corpus_metadata.csv', 'row': 0}, page_content='filename: AdapterFusion_Non-Destructive_Task_Composition_for_Transfer_Learning.pdf\ntitle: AdapterFusion: Non-Destructive Task Composition for Transfer Learning\nabstract: Sequential fine-tuning and multi-task learning are methods aiming to incorporate knowledge from multiple tasks; however, they suffer from catastrophic forgetting and difficulties in dataset balancing. To address these shortcomings, we propose AdapterFusion, a new two stage learning algorithm that leverages knowledge from multiple tasks. First, in the knowledge extraction stage we learn task specific parameters called adapters, that encapsulate the task-specific information. We then combine the adapters in a separate knowledge composition step. We show that by separating the two stages, i.e., knowledge extraction and knowledge composition, the classifier can effectively exploit the representations learned from multiple tasks in a non-de

In [19]:
# html data loading
from langchain_community.document_loaders import UnstructuredHTMLLoader

loader = UnstructuredHTMLLoader("data/Spurious_Correlations.html")
loaded_data = loader.load()
loaded_data

HTTPError: HTTP Error 403: Forbidden

The error HTTPError: HTTP Error 403: Forbidden indicates that the code is trying to access a web resource ("data/Spurious_Correlations.html") and the server is refusing the connection. A 403 error specifically means "Forbidden" – you are not authorized to access that resource.

In [26]:
# pdf data loading
from langchain_community.document_loaders import PyPDFLoader

loader = PyPDFLoader("data/Attention_Is_All_You_Need.pdf")
loaded_data = loader.load_and_split()

In [27]:
loaded_data

[Document(metadata={'source': 'data/Attention_Is_All_You_Need.pdf', 'page': 0}, page_content='Provided proper attribution is provided, Google hereby grants permission to\nreproduce the tables and figures in this paper solely for use in journalistic or\nscholarly works.\nAttention Is All You Need\nAshish Vaswani∗\nGoogle Brain\navaswani@google.com\nNoam Shazeer∗\nGoogle Brain\nnoam@google.com\nNiki Parmar∗\nGoogle Research\nnikip@google.com\nJakob Uszkoreit∗\nGoogle Research\nusz@google.com\nLlion Jones∗\nGoogle Research\nllion@google.com\nAidan N. Gomez∗ †\nUniversity of Toronto\naidan@cs.toronto.edu\nŁukasz Kaiser∗\nGoogle Brain\nlukaszkaiser@google.com\nIllia Polosukhin∗ ‡\nillia.polosukhin@gmail.com\nAbstract\nThe dominant sequence transduction models are based on complex recurrent or\nconvolutional neural networks that include an encoder and a decoder. The best\nperforming models also connect the encoder and decoder through an attention\nmechanism. We propose a new simple network a

In [28]:
loader = PyPDFLoader("data/Attention_Is_All_You_Need.pdf")
loaded_data = loader.load()
loaded_data

[Document(metadata={'source': 'data/Attention_Is_All_You_Need.pdf', 'page': 0}, page_content='Provided proper attribution is provided, Google hereby grants permission to\nreproduce the tables and figures in this paper solely for use in journalistic or\nscholarly works.\nAttention Is All You Need\nAshish Vaswani∗\nGoogle Brain\navaswani@google.com\nNoam Shazeer∗\nGoogle Brain\nnoam@google.com\nNiki Parmar∗\nGoogle Research\nnikip@google.com\nJakob Uszkoreit∗\nGoogle Research\nusz@google.com\nLlion Jones∗\nGoogle Research\nllion@google.com\nAidan N. Gomez∗ †\nUniversity of Toronto\naidan@cs.toronto.edu\nŁukasz Kaiser∗\nGoogle Brain\nlukaszkaiser@google.com\nIllia Polosukhin∗ ‡\nillia.polosukhin@gmail.com\nAbstract\nThe dominant sequence transduction models are based on complex recurrent or\nconvolutional neural networks that include an encoder and a decoder. The best\nperforming models also connect the encoder and decoder through an attention\nmechanism. We propose a new simple network a

In [29]:
# this is the content of the first page
loaded_data[0].page_content

'Provided proper attribution is provided, Google hereby grants permission to\nreproduce the tables and figures in this paper solely for use in journalistic or\nscholarly works.\nAttention Is All You Need\nAshish Vaswani∗\nGoogle Brain\navaswani@google.com\nNoam Shazeer∗\nGoogle Brain\nnoam@google.com\nNiki Parmar∗\nGoogle Research\nnikip@google.com\nJakob Uszkoreit∗\nGoogle Research\nusz@google.com\nLlion Jones∗\nGoogle Research\nllion@google.com\nAidan N. Gomez∗ †\nUniversity of Toronto\naidan@cs.toronto.edu\nŁukasz Kaiser∗\nGoogle Brain\nlukaszkaiser@google.com\nIllia Polosukhin∗ ‡\nillia.polosukhin@gmail.com\nAbstract\nThe dominant sequence transduction models are based on complex recurrent or\nconvolutional neural networks that include an encoder and a decoder. The best\nperforming models also connect the encoder and decoder through an attention\nmechanism. We propose a new simple network architecture, the Transformer,\nbased solely on attention mechanisms, dispensing with recurren

In [30]:
# wikipedia data loading
from langchain_community.document_loaders import WikipediaLoader

loader = WikipediaLoader(query="Tesla", load_max_docs=1)
loaded_data = loader.load()[0].page_content
loaded_data

'Nikola Tesla (; Serbian Cyrillic: Никола Тесла [nǐkola têsla]; 10 July 1856 – 7 January 1943) was a Serbian-American engineer, futurist, and inventor. He is known for his contributions to the design of the modern alternating current (AC) electricity supply system.\nBorn and raised in the Austrian Empire, Tesla first studied engineering and physics in the 1870s without receiving a degree. He then gained practical experience in the early 1880s working in telephony and at Continental Edison in the new electric power industry. In 1884 he immigrated to the United States, where he became a naturalized citizen. He worked for a short time at the Edison Machine Works in New York City before he struck out on his own. With the help of partners to finance and market his ideas, Tesla set up laboratories and companies in New York to develop a range of electrical and mechanical devices. His AC induction motor and related polyphase AC patents, licensed by Westinghouse Electric in 1888, earned him a c

In [33]:
# pass the loaded data to the chat model as context to perform question answering
from langchain_community.document_loaders import WikipediaLoader
from langchain_core.prompts import ChatPromptTemplate

loader = WikipediaLoader(query="Tesla", load_max_docs=1)
loaded_data = loader.load()[0].page_content


chat_template = ChatPromptTemplate.from_messages(
    [
        ("human", "Answer this {question}, here is some extra {context}")
    ]
)

messages = chat_template.format_messages(
    name="Tesla",
    question="Where was he born?",
    context=loaded_data
)

response = chat_model.invoke(messages)
response.content

'Nikola Tesla was born in the village of Smiljan, within the Military Frontier, in the Austrian Empire (present-day Croatia).'

#### Drawbacks:
* Input limit issue (If the number of tokens in the document is more than the context window size of the GPT model, then we will not be able to feed it to the model. Check the details of context window & max output tokens of GPT-3.5 Turbo model: https://platform.openai.com/docs/models/gpt-3.5-turbo)
* Higher Cost (As the number of tokens are more, it will be costly for us. Check the pricing: https://platform.openai.com/docs/pricing)

**Is there any way to overcome these drawbacks?**

Yes, with the help of Retreiver Augmented Generation (RAG).